comms/src/protocol/rpc/test.rs

```diff
@@ -0,0 +1,540 @@
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::{body::Streaming, message::Request, Response, RpcError, RpcServer, RpcStatus, RpcStatusCode};
+use crate::{
+    framing,
+    memsocket::MemorySocket,
+    protocol::{ProtocolEvent, ProtocolId, ProtocolNotification},
+    runtime,
+    runtime::task,
+    test_utils::node_identity::build_node_identity,
+};
+use async_trait::async_trait;
+use futures::{channel::mpsc, stream, SinkExt, StreamExt};
+use std::{io, sync::Arc, time::Duration};
+use tari_shutdown::Shutdown;
+use tari_test_utils::unpack_enum;
+use tokio::{sync::RwLock, time};
+
+#[async_trait]
+// #[tari_rpc(module = generated, protocol_name = "/tari/greeting/1.0")]
+pub trait GreetingRpc: Send + Sync + 'static {
+    // #[rpc(method = 1)]
+    async fn say_hello(&self, request: Request<SayHelloRequest>) -> Result<Response<SayHelloResponse>, RpcStatus>;
+    // #[rpc(method = 2)]
+    async fn return_error(&self, request: Request<()>) -> Result<Response<()>, RpcStatus>;
+    // #[rpc(streaming, method = 3)]
+    async fn get_greetings(&self, request: Request<u32>) -> Result<Streaming<String>, RpcStatus>;
+    // #[rpc(streaming, method = 4)]
+    async fn streaming_error(&self, request: Request<String>) -> Result<Streaming<String>, RpcStatus>;
+    // #[rpc(streaming, method = 5)]
+    async fn streaming_error2(&self, _: Request<()>) -> Result<Streaming<String>, RpcStatus>;
+}
+
+async fn setup_service<T: GreetingRpc>(
+    service: T,
+) -> (
+    mpsc::Sender<ProtocolNotification<MemorySocket>>,
+    task::JoinHandle<Result<(), RpcError>>,
+    Shutdown,
+) {
+    let (notif_tx, notif_rx) = mpsc::channel(1);
+    let shutdown = Shutdown::new();
+    let server_hnd = task::spawn(
+        RpcServer::new()
+            .with_minimum_client_deadline(Duration::from_secs(0))
+            .with_shutdown_signal(shutdown.to_signal())
+            .add_service(generated::server::GreetingService::new(service))
+            .serve(notif_rx),
+    );
+    (notif_tx, server_hnd, shutdown)
+}
+
+async fn setup<T: GreetingRpc>(service: T) -> (MemorySocket, task::JoinHandle<Result<(), RpcError>>, Shutdown) {
+    let (mut notif_tx, server_hnd, shutdown) = setup_service(service).await;
+    let (inbound, outbound) = MemorySocket::new_pair();
+    let node_identity = build_node_identity(Default::default());
+
+    // Notify that a peer wants to speak the greeting RPC protocol
```

Comment:
Looking forward to see how this is plumbed into the current inbound pipeline. This will come after Yamux?

Reply:
Correct (although there is no "inbound pipeline" here), this is why the RPC components are "socket-agnostic". In this test we use a memory socket, but in the real world, the PeerConnection will negotiate and provide a substream. 

The basic idea is that you somehow get hold of a `PeerConnection` (i.e a handle to the underlying connection to a peer - there are a few ways that one would do that depending on the use case) and you would call something like `peer_connection.connect_rpc_client::<MySuperClient>().await`  or `MySuperClient::connect(&mut peer_connection)`.